### ISSUE TITLE
Low performance when typing

### ISSUE THREAD

**Issue by Powersource** [remirror/remirror#1366]
### Summary

Typing in remirror can feel pretty slow compared to typing in e.g. a normal html textarea or in a wysiwyg editor such as https://tiptap.dev/

### Steps to reproduce

Go to https://remirror.io/ and try typing in the editor there and see how quick and snappy it feels. Go to https://tiptap.dev/ and compare. If you have a high performance computer I recommend opening the Chrome dev tools and switching to "Low-end mobile" mode, the difference should be pretty clear using that.

### Expected results

A quick and snappy user experience.

### Actual results

Slow typing, a long time for text to appear (relatively).

**Comment by whawker** [comment]
Interesting, have you seen this slowness is your own editors? Or only on the one displayed on the Remirror website?

If the former, does the speed improve if you remove the floating toolbar?

**Comment by catoenm** [comment]
We're experiencing the same problem. With larger files in Remirror, editing becomes unusable. We don't see this problem when sandboxing Tiptap.

**Comment by Powersource** [comment]
@whawker yes this also happened in our app. What exactly do you mean by the floating toolbar, do you just mean the toolbar? You can see the current iteration of the editor in https://github.com/cobudget-org/cobudget/blob/94a506bd288757244970c5feb46b1b654e58f745/ui/components/Wysiwyg.js , a part of this PR https://github.com/cobudget-org/cobudget/pull/360

**Comment by carlosbaraza** [comment]
@whawker great work on the library. It is really useful.

However, I am struggling to optimise the performance of the library when using multiple extensions and the Toolbar. It seems like the Toolbar is rerendering for every state change, blocking the typing. Do you have any ideas on how to improve performance?

Maybe the problem has to do with the reactivity of the Toolbar telling you whether a command is on or off? Maybe disabling that reactivity could do the trick?

**Comment by carlosbaraza** [comment]
Just noticed that when using the `Remirror` controlled passing the `state` and `onChange`, it has a low performance.

**Comment by carlosbaraza** [comment]
Another trick is to have `useMemo` when embedding your `Editor` component to prevent rendering on every keystroke if you are passing the state to the parent through something like `onChange`

**Comment by nappalm** [comment]
I currently have the same problem, simply using the basic guide 😓 [5-min-tutorial](https://remirror.io/docs/5-min-tutorial)

**Comment by ProTip** [comment]
On the home page it appears to be all React rendering based. TipTap as well.

Both projects make the same architectural decision to trigger component tree re-renders based on the editor state updating by passing the state down the tree for things like controls and etc to respond to.

I have spent quite a bit of time in the past profiling prosemirror and solving for this in a React based project. The solution I went with was to derive editor state facts into a MobX observable on every update. Controls and other components interested in these derivations would consume that observable and only updated when the information they were interested in changes.

Judicious use of Memo is always advisable, but I find it best to avoid as much React rendering logic as possible particularly when dealing with something as latency sensitive as an editor keypress. You want to leave as much CPU time idle as possible in this single-threaded world to mitigate keypress latency on low power devices or during contention(from other areas of the app like processing web request responses and etc).

**Comment by sandzone** [comment]
> Just noticed that when using the `Remirror` controlled passing the `state` and `onChange`, it has a low performance. Another approach if you can is to avoid controlling the component, and just passing the `manager` prop.

uncontrolled editor definitely improves performance... but, there is still a significant drag from nodeviews getting rerendered on every keystroke!
